docs/NFF-0.1-spec.norg

```diff
@@ -0,0 +1,778 @@
+
+*** What is Norg?
+    The _Norg_ syntax is a plain-text file format which aims to be human readable when viewed
+    standalone while also providing a suite of markup utilities for typesetting structured
+    documents. Compared to other plain-text file formats like e.g. Markdown, RST or AsciiDoc, it
+    sets itself apart most notably by following a strict philosophy to abide by the following simple
+    rules:
+    ~ *Consistency:* the syntax should be consistent. Even if you know only a part of the syntax,
+      learning new parts should not be surprising and rather feel predictable and intuitive.
+    ~ *Unambiguity:* the syntax should leave _no_ room for ambiguity. This is especially motivated
+      by the use of [tree-sitter] for the original syntax parser which takes a strict left-to-right
+      parsing approach and only has single-character look-ahead.
+    ~ *[free-form]:* whitespace is _only_ used to delimit tokens but has no other significance! This
+      is probably the most contrasting feature to other plain-text formats which often adhere to the
+      [off-side rule], meaning that the syntax relies on whitespace-indentation to carry meaning.
+
+    [tree-sitter]{https://tree-sitter.github.io/tree-sitter/}
+    [off-side rule]{https://en.wikipedia.org/wiki/Off-side_rule}
+    [Free-form]{https://en.wikipedia.org/wiki/Free-form_language}
+
+** Preliminaries
+   First, we define some basic concepts which will be used in this specification.
+
+
+*** Characters
+    A _Norg_ file is made up of /characters/.
+    A <character> is any Unicode [code point].
+
+    [code point]{https://en.wikipedia.org/wiki/Code_point}
+
+**** Whitespace
+     A {> character} is considered to be *whitespace* if it is any of the following:
+     - A tab `U+0009`
+     - A line feed `U+000A`
+     - A form feed `U+000C`
+     - A carriage return `U+000D`
+     - Any code point in the [Unicode Zs general category] (includes the normal space `U+0020`)
+
+     Any combination of the above is also considered whitespace.
+
+     Tabs are not expanded to spaces and since whitespace has no semantic meaning there is no need
+     to define a default tab stop. However, if a parser must (for implementation reasons) define a
+     tab stop, we suggest to set it to 4 spaces.
+
+     However, the following whitespace combinations are a <line ending>:
+     - A single line feed
+     - A single carriage return
+     - A carriage return immediately followed by a line feed
+
+     ^ Blank lines are irrelevant in _Norg_ because standalone whitespace has no semantic relevance.
+
+     [Unicode Zs general category]{https://www.fileformat.info/info/unicode/category/Zs/list.htm}
+
+**** Punctuation
+     A {> character} is considered to be *punctuation* if it is any of the following:
+     - A standard ASCII punctuation character: |`!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~`|
+
+       +TODO: check whether we want to support every punctuation mark from the `en_US.utf8` locale+
```

Comment:
This may be a bit of a tricky one. I've resorted to looking at the commonmark spec to see how it's done over there:

> An ASCII punctuation character is !, ", #, $, %, &, ', (, ), *, +, ,, -, ., / (U+0021–2F), :, ;, <, =, >, ?, @ (U+003A–0040), [, \, ], ^, _, ` (U+005B–0060), {, |, }, or ~ (U+007B–007E).
>
> A Unicode punctuation character is an ASCII punctuation character or anything in the general Unicode categories Pc, Pd, Pe, Pf, Pi, Po, or Ps.

We may perhaps be interested in supporting a similar (if not the same) range of punctuation chars - these will probably be automatically covered via `std::iswpunct`, which will use the current locale.
The problem I can see arising though is minor inconsistencies or `std::iswpunct` supporting more than what we'd expect.

A possible solution to this could be to enforce `en_US.utf8` as the only natively supported locale - this can be done through `std::setlocale(LC_ALL, "en_US.utf8");` - but then we could run into issues with people using "other" punctuation chars like you said.

We may want to discuss this on discord as it's a pretty difficult thing to get right. The org syntax draft does not specify what punctuation is :(